Horn.fs

```diff
@@ -452,17 +495,110 @@ let hsfTerm
           (hsfFunc definer g)
           (hsfConditionBody definer w c.Semantics) // TODO: keep around Command?
 
+/// <summary>
+///     Constructs a Horn clause for a deferred check, if possible.
+/// </summary>
+let hsfModelDeferredCheck (svars : VarMap) (check : DeferredCheck)
+  : Result<Horn list, Error> =
+    let svarSeq = toVarSeq svars
+
+    let subIteratorInPred iterator f (pred : Func<VIntExpr>) =
+        let fOnIter param =
+            match param with
+            | AVar var when AVar var = iterator -> f var
+            | x -> x
+        { pred with Params = List.map fOnIter pred.Params }
+
+    match check with
+    | NeedsBaseDownclosure (func, reason) ->
+        (* TODO(CaptainHayashi): We're given the func needing downclosure in
+           unflattened form.  This is kind-of messy, as we now have to flatten
+           it again. *)
+        let flatFunc = Starling.Flattener.flattenDView svarSeq [func]
+        let funcPredResult = predOfFunc ensureArith flatFunc
+
+        // TODO(CaptainHayashi): lots of duplication here.
+        let iterator = func.Iterator
+        let iterVarResult =
+            match iterator with
+            | Some x -> ensureArith x
+            | None -> fail (CannotCheckDeferred (check, "malformed iterator"))
+        let funcPredZeroResult =
+            lift2
+                (fun it pred -> subIteratorInPred it (fun _ -> AInt 0L) pred)
+                iterVarResult
+                funcPredResult
+
+        let empPredResult = predOfEmp svars
+
+        let hornResult =
+            lift2
+                (fun zero emp -> Clause (Pred emp, [Pred zero]))
+                funcPredZeroResult
+                empPredResult
+
+        lift
+            (fun h ->
+                [ Comment (sprintf "base downclosure on %s: %s" func.Func.Name reason)
+                  h ])
+            hornResult
+    | NeedsInductiveDownclosure (func, reason) ->
+        // See above for caveats.
+        let flatFunc = Starling.Flattener.flattenDView svarSeq [func]
+        let funcPredResult = predOfFunc ensureArith flatFunc
+
+        let iterator = func.Iterator
+        let iterVarResult =
+            match iterator with
+            | Some x -> ensureArith x
+            | None -> fail (CannotCheckDeferred (check, "malformed iterator"))
+
+        let funcPredSuccResult =
+            lift2
+                (fun it pred -> subIteratorInPred it incVar pred)
+                iterVarResult
+                funcPredResult
+
+        let hornResult =
+            lift3
+                (fun it succ norm ->
+                    Clause (Pred norm, [Ge (it, AInt 0L); Pred succ]))
```

Comment:
My gut feeling is it'd be nicer if the paper used `n+1`, but I'm not sure what you guys think.

@septract @BenSimner ?